### PR TITLE
Refactor live

### PR DESCRIPTION
**Pull request by arj03**
This is a PR on top of #41. Very much WIP as I ripped out live functionality and need to add it back again. But I wanted to try and simplify the code first and then add live again. 

We need to keep the prototype pattern and can't use classes because of [performance](https://stackoverflow.com/questions/3493252/javascript-prototype-operator-performance-saves-memory-but-is-it-faster). So I'm leaving this here to gather input from @staltz while I mull over how to go about adding live again :)

The biggest change here is the introducting of a state machine to (*I think*) clear up a lot of wierd logic with multiple booleans that need to align just right to work.

### CODE REVIEW

File: stream.js [hunk 1]
```diff
@@ -84,19 +93,19 @@ Stream.prototype._writeToSink = function (data) {
 Stream.prototype._handleBlock = function(block) {
   while (true) {
     if (this.sink.paused) return BLOCK_STATE.PAUSED
+
+    // we could send in skip!
```
Comment (arj03) [inline stream.js:97]: I'd still like to see if I can refactor this as well

**Comment by staltz** [comment]
Code looks good but give me some time to carefully scan this logic because it certainly is tricky and important logic.

**Comment by arj03** [comment]
Absolutely, no rush with this. I did not find any new bugs, I ran the [Stress test resume](https://github.com/ssb-ngi-pointer/async-append-only-log/pull/42/commits/27bbe94ca32431c5db982f98f4e2b1cd3053a079) on master as well and it didn't fail.

**Review by staltz — APPROVED** [review]
I tried to mentally translate the `this.hasWritten`, `this.writing`, `this.ended` boolean madness to the new state machine, and it was very confusing, but then I tried to just read the new version of `stream.js` alone and it made a lot of sense, so I feel good in merging this.

**Comment by arj03** [comment]
Agree. It was really confusing before. I feel a lot better with this version. I'll merge this into the other branch and merge into master tonight. Thanks for the review :)

**Pull request merged**